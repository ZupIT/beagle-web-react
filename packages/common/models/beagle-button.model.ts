import { ClickEvent } from '@zup-it/beagle-web'
import { BeagleComponent } from '../types'
import { BeagleDefaultComponent } from './types'


export interface BeagleButtonInterface extends BeagleDefaultComponent, BeagleComponent {
  text: string,
  onPress?: () => void,
  clickAnalyticsEvent?: ClickEvent,
  disabled?: boolean, 
}
