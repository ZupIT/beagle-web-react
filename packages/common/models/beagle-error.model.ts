import { ErrorComponentParams } from '@zup-it/beagle-web'
import { BeagleDefaultComponent } from './types'

export interface BeagleErrorInterface extends BeagleDefaultComponent, ErrorComponentParams {
}
