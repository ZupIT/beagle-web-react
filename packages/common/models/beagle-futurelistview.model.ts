/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { BeagleUIElement } from '@zup-it/beagle-web'
import { BeagleComponent } from '../types'
import { Direction, BeagleDefaultComponent } from './types'

export type NodeType = HTMLElement | null

export interface BeagleFutureListViewInterface extends BeagleDefaultComponent, BeagleComponent {
  direction: Direction,
  dataSource: any[],
  iteratorName?: string,
  onInit?: () => void,
  onScrollEnd?: () => void,
  scrollEndThreshold?: number,
  template: BeagleUIElement,
  useParentScroll?: boolean,
  /* the property "key" is not allowed in React. Since this is not a rule for Beagle, every time
  Beagle receives "key", it transforms it into "_key" */
  _key?: string,
  __suffix__?: string,
}

export interface ScrollInterface {
  direction: Direction,
  onScrollEnd?: () => void,
  scrollEndThreshold: number,
  useParentScroll: boolean,
  elementRef: React.MutableRefObject<HTMLDivElement>,
  hasRendered: boolean,
}
