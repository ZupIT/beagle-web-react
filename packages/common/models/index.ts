/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/
export * from './beagle-image.model'
export * from './beagle-text.model'
export * from './beagle-button.model'
export * from './beagle-listview.model'
export * from './beagle-modal.model'
export * from './beagle-input.model'
export * from './beagle-container.model'
export * from './beagle-touchable.model'
export * from './beagle-tabbar.model'
export * from './beagle-pageview.model'
export * from './beagle-pageindicator.model'
export * from './beagle-lazy.model'
export * from './beagle-simpleform.model'
export * from './beagle-webview.model'
export * from './beagle-futurelistview.model'
export * from './beagle-screen.model'
export * from './beagle-error.model'
export * from './types'
