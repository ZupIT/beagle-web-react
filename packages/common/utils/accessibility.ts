/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { Accessibility } from '../models/accessibility'

type A11YAttr = { qualifiedName: string, value: string }

/* WARNING: If one day Beagle's BFF return more Accessibility props, please add here */
const handlers: Record<keyof Accessibility, (a11y: Accessibility) => A11YAttr | null> = {
  accessibilityLabel: ({ accessibilityLabel }) => 
    accessibilityLabel ? { value: accessibilityLabel || '', qualifiedName: 'aria-label' } : null,
  accessible: () => null,
  isHeader: ({ isHeader }) => isHeader ? { qualifiedName: 'role', value: 'heading' } : null,
}

const notAccessible: Accessibility = { accessible: false }

export const buildAccessibility = (a11y = notAccessible): Record<string, string> => {
  if (!a11y.accessible) return {}

  const keys = Object.keys(a11y) as (keyof Accessibility)[]
  return keys.reduce((result, key) => {
    const attr = handlers[key](a11y)
    return attr ? { ...result, [attr.qualifiedName]: attr.value } : result
  }, {})
}
