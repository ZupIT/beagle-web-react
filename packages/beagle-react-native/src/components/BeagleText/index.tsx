/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import React, { FC } from 'react'
import { BeagleTextInterface, MobileAlignment } from 'common/models'
import { Text, StyleSheet } from 'react-native'

const alignMap: Record<string, MobileAlignment> = {
  auto: 'auto',
  center: 'center',
  left: 'left',
  right: 'right',
}

const BeagleText: FC<BeagleTextInterface> = props => {
  const { text, textColor, alignment, style } = props

  const parsedAlignment = alignment && alignment != 'INHERIT' ?
    alignMap[alignment.toLowerCase()] :
    'auto'
  const styleSheet = StyleSheet.create({
    fromBffStyles: {
      ...style,
    },
    defaultStyles: {
      color: textColor || '#000000',
      textAlign: parsedAlignment,
    },
  })

  return (
    <Text
      style={{
        ...styleSheet.defaultStyles,
        ...styleSheet.fromBffStyles,
      }}>
      {text}
    </Text>
  )
}

export default BeagleText
