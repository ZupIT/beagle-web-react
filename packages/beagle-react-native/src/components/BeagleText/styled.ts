/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import styled from 'styled-components/native'
import { TextAlignment, MobileAlignment } from 'common/models'
import { convertCssStylesToString } from '../utils'

interface StyledTextInterface {
  textColor?: string,
  alignment?: TextAlignment,
  cssStyles?: React.CSSProperties,
}

const alignMap: Record<string, MobileAlignment> = {
  auto: 'auto',
  center: 'center',
  left: 'left',
  right: 'right',
}

export const StyledText = styled.Text<StyledTextInterface>`
color: ${({ textColor }) => textColor ? textColor : '#000000'};
text-align: ${({ alignment }) => alignment && alignment != 'INHERIT' ?
    alignMap[alignment.toLowerCase()] :
    'auto'};
${({ cssStyles }) => cssStyles ? convertCssStylesToString(cssStyles) : ''};
`
