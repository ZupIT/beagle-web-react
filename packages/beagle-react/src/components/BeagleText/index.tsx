/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import React, { FC } from 'react'
import { BeagleTextInterface } from 'common/models'
import { buildAccessibility } from '../../../../common/utils/accessibility'
import withTheme from '../utils/withTheme'
import { StyledText } from './styled'

const BeagleText: FC<BeagleTextInterface> = props => {
  const { text, className, textColor, alignment, style, accessibility } = props
  const a11y = buildAccessibility(accessibility)

  return (
    <StyledText 
      textColor={textColor} 
      alignment={alignment} 
      className={className} 
      style={style}
      {...a11y}>
      {text}
    </StyledText>
  )
}

export default withTheme(BeagleText)
