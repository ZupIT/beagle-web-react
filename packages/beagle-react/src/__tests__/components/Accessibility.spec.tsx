/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

// Link.react.test.js
import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import { configure, mount, shallow } from 'enzyme'
import BeagleText from '../../components/BeagleText'
import { Accessibility } from '../../../../common/models/accessibility'
import { findDOMNode } from 'react-dom'

const ariaLabelText = 'Test Label'
const getAttrsAsArrayFromElement = (element: HTMLElement) =>  [...Array.from(element?.attributes || [])]
const getOnlyAriaBeagleA11YttrsFn = (a: Attr) => a.name.includes('aria-')
const getOnlyRoleBeagleA11YttrsFn = (a: Attr) => (a.name === 'role')
const getBeagleA11YttrsFn = (a: Attr) => getOnlyAriaBeagleA11YttrsFn(a) || getOnlyRoleBeagleA11YttrsFn(a)
const getAttributeValue = (attrs: NamedNodeMap, name: string) => attrs.getNamedItem(name)?.value || ''
const buildAccessibility = 
  (accessible: boolean, accessibilityLabel: string, isHeader: boolean): Accessibility => 
    ({ accessible, accessibilityLabel, isHeader })

configure({ adapter: new Adapter() })

test('Beagle snapshot Accessibility', () => {
  let wrapper = shallow(<BeagleText text='Test' className='test-class'  />)
  expect(wrapper).toMatchSnapshot()
})

// eslint-disable-next-line max-len
test('It should have no "aria-" attributes or "role" attribute when no accessibility is provided or "accessibility.accessible" is false', () => {
  let wrapper = mount(<BeagleText text='Test' className='test-class' />)
  let p = findDOMNode(wrapper.instance()) as HTMLParagraphElement
  
  let attrs = getAttrsAsArrayFromElement(p).filter(getBeagleA11YttrsFn)
  expect(attrs.length).toEqual(0)

  const a11y = buildAccessibility(false, ariaLabelText, true)
  wrapper = mount(<BeagleText text='Test' className='test-class' accessibility={a11y} />)
  p = findDOMNode(wrapper.instance()) as HTMLParagraphElement

  attrs = getAttrsAsArrayFromElement(p).filter(getBeagleA11YttrsFn)
  expect(attrs.length).toEqual(0)
})

it('should set the attribute [role="heading"] when "accessible" and "isHeader" are true', () => {
  const a11y = buildAccessibility(true, '', true)
  const wrapper = mount(<BeagleText text='Test' className='test-class' accessibility={a11y} />)
  const p = wrapper.getDOMNode() as HTMLParagraphElement

  const attrs = getAttrsAsArrayFromElement(p).filter(getOnlyAriaBeagleA11YttrsFn)

  expect(attrs.length).toEqual(0)
  expect(getAttributeValue(p.attributes, 'role')).toEqual('heading')
})

// eslint-disable-next-line max-len
it(`should set the attribute [aria-label="${ariaLabelText}"] when accessibilityLabel is setted`, () => {
  const a11y = buildAccessibility(true, ariaLabelText, false)
  const wrapper = mount(<BeagleText text='Test' className='test-class' accessibility={a11y} />)
  const p = wrapper.getDOMNode() as HTMLParagraphElement

  let attrs = getAttrsAsArrayFromElement(p).filter(getOnlyAriaBeagleA11YttrsFn)
  expect(attrs.length).toBeGreaterThan(0)

  attrs = getAttrsAsArrayFromElement(p).filter(getOnlyRoleBeagleA11YttrsFn)
  expect(attrs.length).toBe(0)

  expect(getAttributeValue(p.attributes, 'aria-label')).toEqual(ariaLabelText)
})

it(`should have role and [aria-label="${ariaLabelText}"] when everything is truthy`, () => {
  const a11y = buildAccessibility(true, ariaLabelText, true)
  const wrapper = mount(<BeagleText text='Test' className='test-class' accessibility={a11y} />)
  const p = wrapper.getDOMNode() as HTMLParagraphElement

  let attrs = getAttrsAsArrayFromElement(p).filter(getOnlyAriaBeagleA11YttrsFn)
  expect(attrs.length).toBeGreaterThan(0)

  attrs = getAttrsAsArrayFromElement(p).filter(getOnlyRoleBeagleA11YttrsFn)
  expect(attrs.length).toBeGreaterThan(0)

  expect(getAttributeValue(p.attributes, 'role')).toEqual('heading')
  expect(getAttributeValue(p.attributes, 'aria-label')).toEqual(ariaLabelText)
})
