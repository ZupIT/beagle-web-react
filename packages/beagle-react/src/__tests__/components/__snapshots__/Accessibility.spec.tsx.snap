// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Beagle snapshot Accessibility 1`] = `
<BeagleText
  className="test-class"
  text="Test"
/>
`;
